sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala

```diff
@@ -91,6 +91,32 @@ case class MapPartitions(
   override def deserializers: Seq[(Expression, Seq[Attribute])] = Seq(deserializer -> child.output)
 }
 
+object MapElements {
+  def apply[T : Encoder, U : Encoder](
+      func: T => U,
+      child: LogicalPlan): MapElements = {
+    MapElements(
+      func.asInstanceOf[Any => Any],
+      encoderFor[T].deserializer,
+      encoderFor[U].namedExpressions,
+      child)
+  }
+}
+
+/**
+ * A relation produced by applying `func` to each element of the `child`.
+ *
+ * @param deserializer used to extract the input to `func` from an input row.
+ * @param serializer use to serialize the output of `func`.
+ */
+case class MapElements(
+    func: Any => Any,
```

Comment:
this should just be AnyRef? then you can use codegen to actually call the java version without wrapping the scala one.